bigframes/ml/compose.py

```diff
@@ -46,6 +48,114 @@
 )
 
 
+CUSTOM_TRANSFORMER_SQL_RX = re.compile(
+    "^(?P<sql>.*)/[*]CT.(?P<id>[A-Z]+[A-Z0-9]*)[(](?P<config>[^*]*)[)][*]/$",
+    re.IGNORECASE,
+)
+
+
+class CustomTransformer(base.BaseTransformer):
```

Comment:
Per my most recent comment, would you be open to renaming this to `SQLScalarColumnTransformer` and accepting `sql` and `columns` as arguments to `__init__`?

This naming would help us align with a similar `bigframes.bigquery.sql_scalar(...)` feature I'd like to add to support folks that want a SQL escape hatch on the `bigframes.pandas.*` APIs.

Reply:
We'll maybe need some additional infrastructure in our compilation to allow nesting multiple `SQLScalarColumnTransformer`s on the same column, but it's possible you've already thought of that :-)

Reply:
> I propose we offer a subset of what you've written here with the following:
> `my_transformer = SQLScalarColumnTransformer(`

So, this is a fixed transformer class to be used inside a ColumnTransformer that wraps all non-"ML." transformations by defining the SQL text with a placeholder for the input column.

This would give us the full flexibility to generate any kind of column transformations in our transformer framework.

When reading in a TRANSFORM_ONLY model, all "ML." lines will be replaced with the corresponding Standard-Transformers and all other lines will be translated into SQLScalarColumnTransformer.

That sound good to me.

Reply:
> renaming this to `SQLScalarColumnTransformer` and accepting `sql` and columns as arguments to `__init__`?
yes, I will rewrite the PR.

Proposal: 
```
__init__(self, sql, columns, output_column_prefix="transformed_", output_columns=None)
```

So, the first syntax would be

```
SQLScalarColumnTransformer(
    "CASE WHEN {0} IS NULL THEN -1 ELSE LENGTH({0}) END",
    columns=["city", "county"],
    output_column_prefix="len_"
)
```

the second syntax would be 


```
SQLScalarColumnTransformer(
    "CASE WHEN {0} IS NULL THEN -1 ELSE LENGTH({0}) END",
    columns=["city", "county"],
    output_columns=["len_city", "county"]
)
```

if not needed I would omit the dtype param.

Reply:
> We'll maybe need some additional infrastructure in our compilation 
> to allow nesting multiple SQLScalarColumnTransformers on the same column, 
> but it's possible you've already thought of that :-)

ColumnTransformer (=BigQuery TRANSFORM_ONLY model) does not support nesting columns.
The only way to apply transformers iteratively on the same column is to create multiple ColumnTransformers and apply them iteratively.

Reply:
`__init__(self, sql, columns, output_column_prefix="transformed_", output_columns=None)`

I was imagining `columns` as the columns that go into the SQL formatter, so one could create features that combine multiple columns. For example:

```python
SQLScalarColumnTransformer(
    "CASE WHEN {0} IS NULL OR {1} IS NULL THEN -1 ELSE LENGTH({0}) + LENGTH({1}) END",
    columns=["city", "county"],
    output_name="len_city_and_county",
)
```

To transform multiple columns, one could create multiple SQLScalarColumnTransformer instances and use them in a Pipeline.

Reply:
As with BQML, all other columns are passed through. So one wouldn't need to name all columns that need to be transformed in one instance.

Reply:
If I anyway have to mention the columns in the constructor, I could immediately replace the placeholders in the sql.
This is anyway possible. So I can access any column name in the sql.
But when the same sql should be applied to different columns the placeholder makes sense.

I am still thinking, that SQLScalarColumnTransformer is only viable inside a ColumnTransformer, right.

Reply:
> I could immediately replace the placeholders in the sql

I was thinking about this more from the bigframes.pandas perspective where the DataFrame column names don't necessarily correspond to a single column, possibly a whole expression. With bigframes.ml, since ML models are more dependent on the correct names for transform and fit steps, this is indeed less of a concern.

I like your proposal, supporting both `output_column_prefix` and `output_columns`, given this bigframes.ml context.

That said, a problem I anticipate is that this approach could cause problems if someone uses the same output column name more than once, such as if they are applying several transforms to the same column with the intention of replacing the original column each time.

Perhaps we can detect when multiple transforms in a pipeline define the one of the same output columns for now? Or does this fail when we try to execute it in BigQuery anyway due to multiple columns with the same name?

Long-term, I think my concerns about nesting transforms will be better addressed as we refactor [bigframes.ml](http://bigframes.ml/) to use more of the bigframes.pandas expression logic.

Reply:
I see a problem reverse engineering the column parameter from the SQL when loading the model.

Just a quick example:

```
col_trans = ColumnTransformer([
    ("lenenc", SQLScalarColumnTransformer(CASE WHEN {0} IS NULL THEN -1 ELSE LENGTH({0}) END), ["city", "county"]),
    ("labelenc", LabelEncoder(), ['county', 'category_name']),
    ...
])
```

After Storing this ColumnTransformer into a TRANSFORM_ONLY model and loading it, it looks different:

col_trans = ColumnTransformer([
    ("lenenc", SQLScalarColumnTransformer(CASE WHEN city IS NULL THEN -1 ELSE LENGTH(city) END), [???]),
    ("lenenc", SQLScalarColumnTransformer(CASE WHEN county IS NULL THEN -1 ELSE LENGTH(county) END), [???]),
    ("labelenc", LabelEncoder(), ['county']),
    ("labelenc", LabelEncoder(), ['category_name']),
    ...
])

Column Transformer uses _parse_from_sql to retrieve the input-column name:

https://github.com/googleapis/python-bigquery-dataframes/blob/4221632e69f1e4866d5db3cc183021ae8f80b245/bigframes/ml/preprocessing.py#L70-L77

But for arbitrary SQL the input column can not be retrieved as simple as `sql[sql.find("(") + 1 : sql.find(")")]`

Reply:
I'd rather not spend too much time trying to restore SQLScalarColumnTransformer as it was originally. Someone could hardcode multiple column names in the SQL string and not even really use the formatting feature.

What if we allowed `columns` to be empty? Alternatively, perhaps a placeholder value could be used for `columns` in such cases, such as `[None]`?

Reply:
Created a first version of the SQLScalarColumnTransformer:

https://github.com/ferenc-hechler/python-bigquery-dataframes/blob/a53b51ccb359f342a0605fb8feb3b32a3fd59dda/bigframes/ml/compose.py#L57-L76